### ISSUE TITLE
Pointer Network

### ISSUE THREAD

**Issue by mataney** [OpenNMT/OpenNMT-py#495]
Hey, I'm looking deeper into the implementation of the Pointer Network and it seem different from other Pointer Network I know ([original paper](https://arxiv.org/pdf/1506.03134.pdf), and other notable papers [1](https://arxiv.org/pdf/1602.06023.pdf), [2](https://arxiv.org/pdf/1704.04368.pdf)

Different specifically in the way we calculate the pointer/generation switch. Here it is done with a non-linear function over the hidden input [link](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/CopyGenerator.py#L94) while in the other resources it is done over the context vector, the decoder state and the decoder input/previous embedding.

Before giving this one a shot and implementing the other designs by myself I was wondering if you have any pointers (couldn't resist) why you chose to implement it this way.

Thanks, Matan.

**Comment by srush** [comment]
No good reason except code arch. Would love a PR. 

I think the right way to do it is to define a general interface for a generator to take each of these components.

**Comment by srush** [comment]
(Oh also just to be clear, our `hidden` is a function of each of the components you mentioned, just indirectly)

**Comment by mataney** [comment]
Hey @srush

I wonder about your second comment, `hidden` is the decoder's output, that is indeed an output of the `context vector`, `decode state` and `decoder input`. 

When I think about it, theoretically this should be enough for the pointer switch to make a decision, but I wonder if passing these 3 separately will have more effect? like they did in the papers I mentioned.

**Comment by srush** [comment]
I could see it having a marginal benefit. It just didn't seem worth it at the time. However, now that the codebase is becoming more general the next step is to abstract attention and generators. So we should be able to implement it exactly.

**Comment by mataney** [comment]
Cool, I hope I'll get to this in the near future, will post a PR.

Another related question, Do you know why, if we decided to point (and not generate) we are using an attention distribution that is learnt specifically for the copy task (`attn['copy']`), and not having just the standard attention and using it to know where to point to?

Thanks again.

**Comment by srush** [comment]
Yeah, we tried both. I think there was a PR to pick which it used. I'll have to check if it got merged.
